src/tests/t_kdb.py

```diff
@@ -336,6 +336,31 @@ def ldap_add(dn, objectclass, attrs=[]):
 realm.kinit('canon', password('canon'))
 realm.kinit('alias', password('canon'), ['-C'])
 
+# Test password history
+def test_pwhist(nhist):
+    output('Testing password history of size ' + `nhist`)
+    princ = 'pwhistprinc' + `nhist`
+    pol = 'pwhistpol'+`nhist`
+    realm.run([kadminl, 'addpol', '-history', `nhist`, pol])
+    realm.run([kadminl, 'addprinc', '-policy', pol, '-nokey', princ])
+    # Test nhist number of passwords and make sure that all previous passwords
+    # fail when trying to be changed to them.
+    cpw = lambda x,y: realm.run([kadminl, 'cpw', '-pw', `x`, princ],
+                                expected_code=y)
+
+    # Creates a combination of password attempts, with (x, 0) and ([0..x], 1)
+    cmb = lambda x: izip(chain([x], islice(count(), x+1)),
+                             chain([0], repeat(1)))
+    # Tries passwords, with one password change that should succeed followed by
+    # all previous passwords that should fail
+    # (0, 0) (0, 1) (1, 0) (0, 1), (1, 1)...
+    list(starmap(cpw, chain.from_iterable(imap(cmb, islice(count(), nhist)))))
+
+    # Test that keys are rotated out correctly
+    cpw(nhist, 0)
+    cpw(0, 0)
+map(test_pwhist, imap(pow, repeat(2), range(5)))
```

Comment:
This treat is specifically designed to test that the key rotation is working right. I had originally written it because I had some concerns over that functionality specifically, and I saw no reason to not include it after that. 3 password histories won't really be enough for that, and anything beyond three will have a lot of commands which would be better to put in a loop. I could add basic history tests so any issues that doing have to do with the rotation will show up in something far easier to read, but I would hesitate to remove this test. I can drop it down to 8 tests which are fairly quick as a compromise.